libraries/opensk/src/api/persist.rs

```diff
@@ -0,0 +1,367 @@
+// Copyright 2024 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+mod keys;
+
+use crate::ctap::status_code::{Ctap2StatusCode, CtapResult};
+use crate::ctap::PIN_AUTH_LENGTH;
+use alloc::boxed::Box;
+use alloc::vec::Vec;
+use core::cmp;
+
+pub type PersistIter<'a> = Box<dyn Iterator<Item = CtapResult<usize>> + 'a>;
+pub type PersistCredentialIter<'a> = Box<dyn Iterator<Item = CtapResult<(usize, Vec<u8>)>> + 'a>;
+
+/// Stores data that persists across reboots.
+///
+/// This trait might get appended to with new versions of CTAP.
+///
+/// To implement this trait, you have 2 options:
+/// - Implement all high level functions with default implementations,
+///   calling `unimplemented!` in the key-value accessors.
+///   When we update this trait in a new version, OpenSK will panic when calling any new functions.
+/// - Implement the key-value accessors, and special case as many default implemented high level
+///   functions as desired.
+///   When the trait gets extended, new features will silently work.
+///   Credentials still need keys to be identified by.
+pub trait Persist {
+    /// Retrieves the value for a given key.
```

Comment:
Since we specify the maximum length below, shouldn't we also specify the maximum key here?

Note that we could also not specify anything, then implementation just return an error if the argument is not supported.

Reply:
Documentation improved, more explicit on trait level.